tests/integration/test_elle.py

```diff
@@ -0,0 +1,293 @@
+        "{}00000003".format(cluster2_dbid).encode(), cluster2.node(3).address,
+    ) == b'OK'
+
+    cluster1.wait_for_unanimity()
+    cluster2.wait_for_unanimity()
+
+    ops = worker.generate_ops(["{test}key1"])
+    worker.do_ops(ops)
+
+    assert cluster1.execute(
+        'migrate', '', '', '', '', '', 'keys',
+        "{test}key1", "{test}key2", "{test}key3", "{test}key4",
+    ) == b'OK'
+
+    cluster1.wait_for_unanimity()
+    cluster2.wait_for_unanimity()
+
+    ops = worker.generate_ops(["{test}key1"])
+    worker.do_ops(ops)
+
+    assert cluster1.execute(
+        'RAFT.SHARDGROUP', 'REPLACE',
+        '2',
+
+        cluster1_dbid,
+        2, 3,
+        0, slot-1, 1,  0,
+        slot+1, 16383, 1, 0,
+        "{}00000001".format(cluster1_dbid).encode(), cluster1.node(1).address,
+        "{}00000002".format(cluster1_dbid).encode(), cluster1.node(2).address,
+        "{}00000003".format(cluster1_dbid).encode(), cluster1.node(3).address,
+
+        cluster2_dbid,
+        1, 3,
+        slot, slot, 1, 0,
+        "{}00000001".format(cluster2_dbid).encode(), cluster2.node(1).address,
+        "{}00000002".format(cluster2_dbid).encode(), cluster2.node(2).address,
+        "{}00000003".format(cluster2_dbid).encode(), cluster2.node(3).address,
+    ) == b'OK'
+
+    assert cluster2.execute(
+        'RAFT.SHARDGROUP', 'REPLACE',
+        '2',
+
+        cluster1_dbid,
+        2, 3,
+        0, slot-1, 1,  0,
+        slot+1, 16383, 1, 0,
+        "{}00000001".format(cluster1_dbid).encode(), cluster1.node(1).address,
+        "{}00000002".format(cluster1_dbid).encode(), cluster1.node(2).address,
+        "{}00000003".format(cluster1_dbid).encode(), cluster1.node(3).address,
+
+        cluster2_dbid,
+        1, 3,
+        slot, slot, 1, 0,
+        "{}00000001".format(cluster2_dbid).encode(), cluster2.node(1).address,
+        "{}00000002".format(cluster2_dbid).encode(), cluster2.node(2).address,
+        "{}00000003".format(cluster2_dbid).encode(), cluster2.node(3).address,
+    ) == b'OK'
+
+    cluster1.wait_for_unanimity()
+    cluster2.wait_for_unanimity()
+
+    ops = worker.generate_ops(["{test}key1"])
+    worker.do_ops(ops)
+
+    assert cluster1.leader_node().raft_debug_exec("lrange", "{test}key1", 0, -1) == []
+    val = cluster2.leader_node().raft_debug_exec("lrange", "{test}key1", 0, -1)
+    assert type(val) is list
+    assert len(val) == 4
+
+
+@pytest.mark.key_hash_tag("test")
+@pytest.mark.num_elle_keys(5)
+def test_elle_migrating(request, cluster_factory):
+    cluster1 = cluster_factory().create(3, raft_args={
+        'sharding': 'yes',
+        'external-sharding': 'yes'})
+    cluster2 = cluster_factory().create(3, raft_args={
+        'sharding': 'yes',
+        'external-sharding': 'yes'})
+
+    cluster1_dbid = cluster1.leader_node().info()["raft_dbid"]
+    cluster2_dbid = cluster2.leader_node().info()["raft_dbid"]
+
+    slot = -1
+    marker = request.node.get_closest_marker("key_hash_tag")
```

Comment:
This prevents making errors but I don't think it is sustainable if we want to reuse tests without polluting it. 

Initially, I thought we would have 16k keys, one per slot. Then, test would run by picking keys randomly on each operation. So, we wouldn't need to specify anything. Maybe this is not so efficient for short running tests, also not sure if Elle is efficient with lots of keys. Manually marking keys make sense but I think we need to extend this to multi slot in future.

Reply:
I believe marks can take arrays, so we should be able to extend it with that (and have a different mark if we want to make sure it touches all slots)

Reply:
My point was we should be able to write tests unaware of Elle as we discussed. So, it means, tests should not do things like `key_hash_slot()` or `get_closest_marker()` etc. This is not the suggested way to write tests right? 

For certain tests, we may want to let Elle know the key prefix that is used in the test. So, it is okay to have a tag `@pytest.mark.key_hash_tag("{test}")`. 

Anyway, we can revisit this after exploring "one key per slot" setup. I think it should be the default. Then, if we see `@pytest.mark.key_hash_tag("{test}")` marker, Elle can operate with that slot / slots only.

Reply:
i don't see a good way to get rid of key_hash_slot() in this test, I can get rid of the marker calls though.

the idea would be that the instead of doing a key_hash_slot() on the value we get from the marker, we do a hash_key_slot on a literal (which is also set as the marker).  Therefore, it be `slot = key_hash_slot("test")`.

the test itself would validate the that key(s) used by the test itself was/were migrated, and not rely on the cheating "count" mechanism, which would remove the need for the conditional at the bottom.

thoughts?

Reply:
Ok, `key_hash_slot()` is probably better/required. Just marker calls are a bit ugly inside a test. It'd be nice to avoid them.